neurom/analysis/dendrogram.py

```diff
@@ -0,0 +1,256 @@
+
+import numpy as np
+import sys
+
+
+def _total_rectangles(tree):
+    '''
+    Calculate the total number of segments that are required
+    for the dendrogram. There is a vertical line for each segment
+    and two horizontal line at each branching point
+    '''
+    return n_segments(tree) + n_bifurcations(tree) * 2
+
+
+def _n_rectangles(obj):
+    '''
+    Calculate the total number of rectangles with respect to
+    the type of the object
+    '''
+    if isinstance(obj, Tree):
+
+        return _total_rectangles(obj)
+
+    elif isinstance(obj, Neuron):
+
+        return sum([_total_rectangles(neu) for neu in obj.neurites])
+
+    else:
+
+        return 0
+
+
+def _vertical_segment(old_offs, new_offs, spacing, radii):
+    '''Vertices fo a vertical rectangle
+    '''
+    return np.array(((new_offs[0] - radii[0], old_offs[1] + spacing[1]),
+                     (new_offs[0] - radii[1], new_offs[1]),
+                     (new_offs[0] + radii[1], new_offs[1]),
+                     (new_offs[0] + radii[0], old_offs[1] + spacing[1])))
+
+
+def _horizontal_segment(old_offs, new_offs, spacing, diameter):
+    '''Vertices of a horizontal rectangle
+    '''
+    return np.array(((old_offs[0], old_offs[1] + spacing[1]),
+                     (new_offs[0], old_offs[1] + spacing[1]),
+                     (new_offs[0], old_offs[1] + spacing[1] - diameter),
+                     (old_offs[0], old_offs[1] + spacing[1] - diameter)))
+
+
+def _spacingx(node, max_dims, xoffset, xspace):
+    '''Determine the spacing of the current node depending on the number
+       of the leaves of the tree
+    '''
+    x_spacing = n_terminations(node) * xspace
+
+    if x_spacing > max_dims[0]:
+        max_dims[0] = x_spacing
+
+    return xoffset - x_spacing / 2.
+
+
+def _update_offsets(start_x, spacing, terminations, offsets, length):
+    '''Update the offsets
+    '''
+    return (start_x + spacing[0] * terminations / 2.,
+            offsets[1] + spacing[1] * 2. + length)
+
+
+class Dendrogram(object):
+    '''Dendrogram
+    '''
+
+    def __init__(self, obj):
+        '''Create dendrogram
+        '''
+
+        # input object, tree, or neuron
+        self._obj = obj
+
+        # counter/index for the storage of the rectangles.
+        # it is updated recursively
+        self._n = 0
+
+        self.scale = 1.
+
+        # the maximum lengths in x and y that is occupied
+        # by a neurite. It is updated recursively.
+        self._max_dims = [0., 0.]
+
+        # stores indices that refer to the _rectangles array
+        # for each neurite
+        self._groups = []
+
+        # dims store the max dimensions for each neurite
+        # essential for the displacement in the plotting
+        self._dims = []
+
+        # initialize the number of rectangles
+        self._rectangles = np.zeros([_n_rectangles(self._obj), 4, 2])
+
+    def generate(self):
+        '''Generate dendrogram
+        '''
+
+        sys.setrecursionlimit(100000)
```

Comment:
We have to find a way of avoiding hitting large recursion depths. In the meantime, you should at lease make sure this function leaves `sys` like it found it, i.e. make sure the limit is set to what it was originally. You could implement this in a decorator to make it more transparent.